boards: arm: mps2_an521: force secure firmware image by default

ioannisg opened the pull request:
In order to increase code coverage, we force building a Secure
Firmware image by default (i.e. with option
CONFIG_TRUSTED_EXECUTION_SECURE set), when building for
mps2_an521 board. CONFIG_TRUSTED_EXECUTION_SECURE enables
compiling-in all TrustZone-related code in the tree, that is,
all ARM-specific code inside #ifdef CONFIG_ARM_SECURE_FIRMWARE.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Closes #16148

ioannisg committed:
boards: arm: mps2_an521: force secure firmware image by default

In order to increase code coverage, we force building a Secure
Firmware image by default (i.e. with option
CONFIG_TRUSTED_EXECUTION_SECURE set), when building for
mps2_an521 board. CONFIG_TRUSTED_EXECUTION_SECURE enables
compiling-in all TrustZone-related code in the tree, that is,
all ARM-specific code inside #ifdef CONFIG_ARM_SECURE_FIRMWARE.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@galak do you think we can get this in?

microbuilder reviewed (approved):
lgtm, but hopefully @karl-zh can give some feedback on this based on the work in #16450, which configures the Zephyr image to run in the non-secure environment, and builds a secure binary using the upstream TF-M repository, talking to the secure services and environment via IPC.

ioannisg committed:
arch: arm: BusFault, NMI, and HardFault in Secure state when in test

This commit enables the option to route the BusFault,
HardFault, and NMI exceptions in Secure state, when
building for Cortex-M CPUs with ARM_SECURE_FIRMWARE=y.
This allows the various test to utilize BusFault,
HardFault and NMI exceptions during testing.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
Currently on top of #18082. The patch is the top 2 commits.

@microbuilder @galak @andrewboie can we try to get that in for 2.0?

In brief:
- the top commit forces` CONFIG_TRUSTED_EXECUTION_SECURE=y` when building for `mps2_an521`. So it forces Secure mode, which is, anyways, the default mode for TrustZone-M CPUs; all the patch does, is to compile-in all code under #ifdef CONFIG_ARM_SECURE_FIRMWARE, so we get this code built, which, otherwise is a _dead_ code, cause we have no samples, yet, for Secure ARM code. So this is about increasing code coverage
- there's also a minor tweak in the second commit, @galak @microbuilder; normally, non-banked faults (BUS, NMI, etc.) are routed to non-secure firmware. This aligns well with TF-M's view of running full Zephyr in non-secure domain (so it would need to utilize all kinds of CPU exceptions). But that would break some tests in the tree that actually generate NMIs and BusFaults (e.g. to test them, such as arch/arm/arm_runtime_nmi suite). So I added a Kconfig option to have this "routing" configurable; and _not_ route these exceptions to Non-Secure if we are in test mode (that is, keep the default behavior in production mode)

ioannisg commented:
FYI @karl-zh
